src/cysignals/signals.pxd

```diff
@@ -51,6 +57,28 @@ cdef inline void cython_check_exception() nogil except *:
     pass
 
 
+cdef void verify_exc_value()
+
+cdef inline PyObject* interrupt_occurred():
+    """
+    Borrowed reference to the exception which is currently being
+    propagated from cysignals. If there is no exception or if we
+    are done handling the exception, return ``NULL``.
+
+    This is meant for Cython code to check whether objects may be in
+    an invalid state. Typically, this would be used in an ``except``
+    or ``finally`` block or in ``__dealloc__``.
+
+    The implementation is based on reference counting to check whether
+    the exception has been deleted. This means that it will break if the
+    exception is stored somewhere or if the exception is involved in a
+    reference cycle.
```

Comment:
Naming: "interrupt_occurred" is a bit confusing; from the machine level yes it's an interrupt, but in the context of signal handling this naming could be confusing with SIGINT.  It also doesn't have the `sig_` prefix of other Cysignals functions.  So I was thinking just something like `sig_occurred()` indicating that Cysignals is handling an exception (somewhat analogous to `PyErr_Occurred`).

Should it really be considered "broken" in that case?  Really it just means a change in semantics--this doesn't *necessarily* mean that the returned exception is currently being handled (though it probably is in most cases).  Instead it's just returning the last exception raised by Cysignals that still has references to it (whether because it's being handled, or was stored in a variable for some reason).

To this end it might also be useful to have a `sig_clear()` to manually do `Py_XDECREF(cysigs.exc_value); cysigs.exc_value = NULL`.  Then, even if we did want to store the exception in a variable, for example, a wary user could do something like:

```
err = sig_occurred()
sig_clear()
... do stuff with err ...
```

Does that make any sense or am I off base?

Reply:
> Should it really be considered "broken" in that case?

Well, the ideal semantic of `sig_occurred` is: return the exception from cysignals if it has not been handled yet. Of course, now it comes down to defining what "handled" means, but intuitively it means that the `try/except/finally` block containing `sig_on` has been exited.

Reply:
A more likely scenario to add a reference to the exception would be
```
typ, val, tb = sys.exc_info()
... do stuff with typ, val, tb ...
```
i.e. the code doing something with the exception is most likely unaware of cysignals.